examples/tv-app/linux/AppImpl.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file Contains shell commands for a commissionee (eg. end device) related to commissioning.
+ */
+
+#include "AppImpl.h"
+
+#include <app-common/zap-generated/attribute-id.h>
+#include <app-common/zap-generated/cluster-id.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/Command.h>
+#include <app/server/Dnssd.h>
+#include <app/server/Server.h>
+#include <app/util/af.h>
+#include <cstdio>
+#include <inttypes.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/shell/Commands.h>
+#include <lib/shell/Engine.h>
+#include <lib/shell/commands/Help.h>
+#include <lib/support/CHIPArgParser.hpp>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/ZclString.h>
+#include <platform/CHIPDeviceLayer.h>
+
+using namespace chip;
+using namespace chip::AppPlatform;
+
+#if CHIP_DEVICE_CONFIG_APP_PLATFORM_ENABLED
+
+namespace chip {
+namespace AppPlatform {
+
+// BEGIN DYNAMIC ENDPOINTS
+// =================================================================================
+
+static const int kNameSize = 32;
+
+// Current ZCL implementation of Struct uses a max-size array of 254 bytes
+static const int kDescriptorAttributeArraySize = 254;
+static const int kFixedLabelAttributeArraySize = 254;
+
+// Device types for dynamic endpoints: TODO Need a generated file from ZAP to define these!
+// (taken from chip-devices.xml)
+#define DEVICE_TYPE_CONTENT_APP 0x0024
+
+// ---------------------------------------------------------------------------
+//
+// CONTENT APP ENDPOINT: contains the following clusters:
+//   - Descriptor
+//   - Application Basic
+
+// Declare Descriptor cluster attributes
+DECLARE_DYNAMIC_ATTRIBUTE_LIST_BEGIN(descriptorAttrs)
+DECLARE_DYNAMIC_ATTRIBUTE(ZCL_DEVICE_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0),     /* device list */
+    DECLARE_DYNAMIC_ATTRIBUTE(ZCL_SERVER_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0), /* server list */
+    DECLARE_DYNAMIC_ATTRIBUTE(ZCL_CLIENT_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0), /* client list */
+    DECLARE_DYNAMIC_ATTRIBUTE(ZCL_PARTS_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0),  /* parts list */
```

Comment:
This isn't right, at least not if we want reading these to actually work.  To make it work, these should be declared as external attributes, size 0, and an AttributeAccessInterface should be implemented to actually produce the values.

I assume this was implemented by copying the bridge code (which is broken and needs to be fixed)?  Was this tested?  Because I expect that all those lists either error out or return a 0-length list depending on what data is in there....

Same for the other ARRAY attributes here.

Reply:
This code was based upon bridge and the original purpose of these declarations was to make sure the dynamic endpoint ember code allows these attributes to be read, however the actual reading is now done by AttributeAccessInterface.